docs/source/developers/python.rst

```diff
@@ -586,6 +586,86 @@ Caveats
 
 The Plasma component is not supported on Windows.
 
+Relevant components and environment variables
+=============================================
+
+List of relevant Arrow CMake flags and corresponding environment variables
+to be used when building PyArrow are:
+
+.. list-table::
+   :widths: 30 30
+   :header-rows: 1
+
+   * - CMAKE_BUILD_TYPE
+     - PYARROW_BUILD_TYPE (release, dbug or relwithdebinfo)
+   * - ARROW_GCS
+     - PYARROW_WITH_GCS
+   * - ARROW_S3
+     - PYARROW_WITH_S3
+   * - ARROW_HDFS
+     - PYARROW_WITH_HDFS
+   * - ARROW_CUDA
+     - PYARROW_WITH_CUDA
+   * - ARROW_SUBSTRAIT
+     - PYARROW_WITH_SUBSTRAIT
+   * - ARROW_FLIGHT
+     - PYARROW_WITH_FLIGHT
+   * - ARROW_DATASET
+     - PYARROW_WITH_DATASET
+   * - ARROW_PARQUET
+     - PYARROW_WITH_PARQUET
+   * - PARQUET_REQUIRE_ENCRYPTION
+     - PYARROW_WITH_PARQUET_ENCRYPTION
+   * - ARROW_PLASMA
+     - PYARROW_WITH_PLASMA
+   * - ARROW_TENSORFLOW
+     - PYARROW_WITH_TENSORFLOW
+   * - ARROW_ORC
+     - PYARROW_WITH_ORC
+   * - ARROW_GANDIVA
+     - PYARROW_WITH_GANDIVA
+
+List of relevant environment variables that can also be used to build
+PyArrow are:
+
+.. list-table::
+   :widths: 30 30
+   :header-rows: 1
+
+   * - PyArrow environment variable
+     - Description
+   * - PYARROW_CMAKE_GENERATOR
+     - Example: 'Visual Studio 15 2017 Win64'
+   * - PYARROW_CMAKE_OPTIONS
+     - Extra CMake and Arrow options (ex. ``"-DARROW_SIMD_LEVEL"``,
+       ``"-DCMAKE_OSX_ARCHITECTURES"``)
+   * - PYARROW_BOOST_NAMESPACE
+     - Name of install folder for boost (if different from 'boost')
+   * - PYARROW_CXXFLAGS
+     - Extra cxx flags
+   * - PYARROW_WITH_STATIC_PARQUET
+     - Rely on parquet shared libraries where relevant, default OFF
+   * - PYARROW_WITH_STATIC_BOOST
+     - Rely on Boost shared libraries on linking static parquet, default OFF
```

Comment:
There is some inconsistency if I am reading the code correctly. 

`PYARROW_WITH_STATIC_PARQUET` and `PYARROW_WITH_STATIC_BOOST` are set to `false`/`OFF` (`0`) in setup.py:
https://github.com/apache/arrow/blob/e69a53e9ee8a1c14b9556c1009c12903d89a594b/python/setup.py#L174-L179

but the corresponding flags are set to `ON` in CMakeLists.txt:
https://github.com/apache/arrow/blob/e69a53e9ee8a1c14b9556c1009c12903d89a594b/python/setup.py#L309-L312
https://github.com/apache/arrow/blob/e69a53e9ee8a1c14b9556c1009c12903d89a594b/python/CMakeLists.txt#L119-L123

Will create an issue for it if I am not mistaken.

Reply:
Because `setup.py` uses `STATIC` (not `SHARED`) and `CMakeLists.txt` uses `SHARED` (not `STATIC`).
(We can change `CMakeLists.txt` to use `STATIC` to align names in `setup.py` and `CMakeLists.txt` for easy to understand.)

We can remove `PYARROW_WITH_STATIC_BOOST` because we don't use Boost now.

Reply:
Sorry if the questions are silly:
But we can link dynamically to Arrow C++ libs in python I think? And it is also a bit confusing for me that setup.py and CMake that is run from setup.py define different ways to link to the libraries.

Yes, it would be much more clear if this could be aligned or at least I can try to understand and add a sentence or two about this to the docs.

Reply:
> But we can link dynamically to Arrow C++ libs in python I think?

Correct.

> setup.py and CMake that is run from setup.py define different ways to link to the libraries.

Does it mean that `setup.py` says "static or not" and `CMake` says "shared or not"?


Ah, wait. We may be able to remove `PYARROW_WITH_STATIC_PARQUET` too like `PYARROW_WITH_STATIC_BOOST` with the following change:

```diff
diff --git a/python/CMakeLists.txt b/python/CMakeLists.txt
index 8f846348f3..2e62dfb0e5 100644
--- a/python/CMakeLists.txt
+++ b/python/CMakeLists.txt
@@ -586,7 +586,7 @@ if(PYARROW_BUILD_PARQUET)
             DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
   endif()
 
-  if(PYARROW_PARQUET_USE_SHARED)
+  if(ARROW_BUILD_SHARED)
     if(PYARROW_BUNDLE_ARROW_CPP)
       bundle_arrow_lib(${PARQUET_SHARED_LIB} SO_VERSION ${PARQUET_SO_VERSION})
       if(MSVC)

```

Reply:
Great! Will include the change and test it locally.